kernel/parray.ml

```diff
@@ -164,38 +168,22 @@ let copy p =
   | Array (t, def) -> ref (Array (UArray.copy t, def))
   | Updated _ -> assert false
 
-let reroot t =
-  let _ = reroot t in
-  t
-
 let map f p =
-  let p = reroot p in
-  match !p with
-  | Array (t,def) ->
-    let t = uinit (UArray.length t) (fun i -> f (UArray.unsafe_get t i)) in
-    ref (Array (t, f def))
-  | Updated _ -> assert false
+  let t = uinit (length_int p) (fun i -> f (unsafe_get p i)) in
```

Comment:
I don't think either that we can avoid checking for rerooting. But the first time it happens, we could just blindly copy the array and resume. That way, we are guaranteed to suffer at most one reroot during conversion.

My fear is that, on some examples, we will end up rerooting like crazy otherwise. Indeed, the examples might seem artificial, but I would not be surprised if they are representative of the access patterns you would get for an operation on matrices (implemented as arrays of array). For example, it would be quite painful if the result of an O(n^3) Gaussian elimination were to be checked for convertibility in O(n^5).

Reply:
Folding is already O(n), what do we lose by copying the array unconditionally at the beginning of the function?

Reply:
I was hoping we could avoid trashing the memory. But since this is lazy conversion we are talking about, memory is already trashed anyway. So, yes, I guess we could always copy the array rather than do it lazily.